Add config tests

cknv opened the pull request:
As mentioned in #102 there was a need for some more test coverage of the config class, which seems like a fairly lowing hanging fruit, possibly low enough that I could manage to pick it.

I used the latest codecov report of the class to see what lines was missing coverage and tried to target those as best I could. I went with the naive way of just asserting what the code returned, so it is not impossible (perhaps likely, even) that I am mistaken about the behaviour of the class.

Locally I have some failing tests in the `test_statreload.py` file, but as I have yet to make any changes to the code itself, I suspect there where there when I cloned the repo. I could have a look at what those are, but I have barely looked past the config class.

cknv committed:
Add tests for most of Config class

cknv committed:
Add ssl config test

cknv commented on tests/test_config.py:
```diff
@@ -0,0 +1,75 @@
+    config = Config(app=asgi_app, ssl_certfile=certfile.name, ssl_keyfile=keyfile.name)
+    config.load()
+
+    assert bool(config.is_ssl) is True
```
The `bool` call should be unnecessary here, but the `.is_ssl` property is defined as such:

```python
@property
def is_ssl(self) -> bool:
    return self.ssl_keyfile or self.ssl_certfile
```
Which of course evaluates as a boolean, but isn't _technically_ a boolean. I could move the bool coercion there if it's appropriate?

jlaine replied:
I'd say fix the problem at the root, `is_ssl` should return an actual boolean, not a truthy / falsy value.

```python
@property
def is_ssl(self) -> bool:
    return bool(self.ssl_keyfile or self.ssl_certfile)
```

and the test should be:

```python
assert config.is_ssl is True
```

I think mypy usually catches these kinds of problems

cknv replied:
Yeah, mypy would probably catch that.

I lean to actually fixing it too, but decided to keep it minimal at first and only add tests.

It's a tiny addition, so I should have time soon to make the change.

cknv committed:
Fix Config.is_ssl property return type

cknv commented:
I have rebased onto the latest master which should make the test suite pass.

cknv commented:
I see the previously failing tests was masking some actual problems, and have a bit more work cut out for me with regards to dealing with python 3.5 and the permissions of the certificates on windows.

cknv committed:
Fix string app import error for python 3.5

cknv commented:
The cert + windows problems seems to stem from the fact that `NamedTemporaryFile` is not able to be re-opened on Windows NT and later (or so is my understanding of the [docs](https://docs.python.org/3.7/library/tempfile.html#tempfile.NamedTemporaryFile) at least) it _shouldn't_ be too hard to fix that too.

cknv committed:
Change cert and key fixture to pytest tmp files

Using NamedTemporaryFile seems to cause problems on Windows builds
as they are not able to be reopened later on, instead it is
possible to use the pytest fixture tmp_path to build more looking
tmp files, and just return the filename, which works fine as only
the names was being used anyway.

cknv committed:
Convert paths to strings before use

cknv committed:
Blacken conftest

cknv commented:
The windows build is still failing, this time when getting `"/"` of `test_request_logging`.

Judging from the [logs on travis](https://travis-ci.org/encode/uvicorn/jobs/577496079#L169-L171), there is an extra task spawned onto the loop for some reason and the mock loop only runs a single coroutine for those tests. It started happening after the change from `NamedTemporaryFile` to using `tmp_path` to just build paths with.

I don't have a windows machine at hand to replicate this with, so if someone can see the connection between the changes I did and the test, I would be very happy to continue, but currently I am just hitting the wall here.

jlaine commented:
CI failures in `test_request_logging` may be a symptom of what I reported in #406 : this test is subject to "pollution" from exceptions raised by other tests. I'm starting to think that clearing the log capture at the beginning of this test would be a good idea.

cknv commented:
Oh yeah @jlaine that looks very similar, and indeed I can actually see a similar log line being spewed out when running the tests locally, although that does not make it fail for some reason (running Arch locally here with Python 3.7.4 and uvloop 0.13.0).
Could it be that _maybe_ its actually some stray asyncio tasks that have been left behind by a test only to be destroyed in another ? It feels a bit like wild speculation, but I cannot shake the thought.

jlaine commented:
I have submitted a tentative patch in #418. AFAICT the issue is that if the WSGI task raises an exception, the "sender" task (in charge of sending out response messages) is left dangling.

lovelydinosaur commented:
Okay, probably worth merging master into this PR, and seeing where that gets us, right?

cknv committed:
Merge branch 'master' into config-tests

* master:
  Relax uvloop version requirement to 0.*
  If the WSGI task blows up, allow the sender task to shut down

cknv commented:
That seemed to do the trick. Very nice @jlaine!

Now that it is actually passing tests, we can

lovelydinosaur commented:
Lovely stuff!
